extensions/amp-analytics/0.1/visibility-impl.js

```diff
@@ -370,22 +372,54 @@ export class Visibility {
 
     const visible = change.intersectionRatio * 100;
     for (let c = listeners.length - 1; c >= 0; c--) {
-      const shouldBeVisible = !!listeners[c]['shouldBeVisible'];
-      if (this.updateCounters_(visible, listeners[c], shouldBeVisible)
-          && this.viewer_.isVisible() == shouldBeVisible) {
-        this.prepareStateForCallback_(
-            listeners[c]['state'], change.boundingClientRect);
-        listeners[c].callback(listeners[c]['state']);
+      const listener = listeners[c];
+      const shouldBeVisible = !!listener.shouldBeVisible;
+      const config = listener.config;
+      const state = listener.state;
+      state[LAST_CHANGE_ENTRY] = change;
```

Comment:
I was following the original code. This is my understanding: 

`state` is the final map that made available for URL substitution, so in that sense, it makes sense to make it a amp.

Meanwhile, `state` is also used to save temporary state data, which will eventually be deleted before passing for URL substitution.

What we can potentially improve for readability is to split the 2 usages. I.e., use a type defined object to store internal state, and convert it to a new map when sending for URL substitution. 

But I'd rather do it in a separate PR, as that'll touch the V1 code. Doing it here will made the current logic change unreviewable.

What you think?